### DISCUSSION TITLE
What's the best way to test a tui application?

### DISCUSSION THREAD

**Discussion by asomers-ax** [fdehau/tui-rs#509] [Q&A]
Tui let me quickly build two graphical applications.  That's great!  But I haven't figured out how to automate testing.  Ideally I'd like tests which assert that canned input produces expected output given a terminal of a certain size.  But tui expects stdin to be a console, and the output, when redirected to a file, is full of unintelliglble terminal control characters.  Does anybody have a good strategy for testing a tui application?

**Comment by fdehau** [comment]
Have you looked how `tui-rs` "integration" tests are written ? We are using `TestBackend` to check the output in the intermediate buffers. See [this test](https://github.com/fdehau/tui-rs/blob/a68e38e59e6735c0a99139303b1609669d2c38da/tests/widgets_block.rs#L12) for an example. This is not exactly what will be written to the terminal (the crate outputs the minimal diff between frames) but it should be enough to check what your UI will look like.
